contracts/integrations/test/framework/assertions/fillOrder.ts

```diff
@@ -0,0 +1,90 @@
+import { ERC20TokenEvents, ERC20TokenTransferEventArgs } from '@0x/contracts-erc20';
+import { ExchangeEvents, ExchangeFillEventArgs } from '@0x/contracts-exchange';
+import { constants, expect, orderHashUtils, verifyEvents } from '@0x/contracts-test-utils';
+import { FillResults, Order } from '@0x/types';
+import { BigNumber, logUtils } from '@0x/utils';
+import { TransactionReceiptWithDecodedLogs } from 'ethereum-types';
+import * as _ from 'lodash';
+
+import { DeploymentManager } from '../deployment_manager';
+
+import { FunctionArguments, FunctionAssertion, FunctionResult } from './function_assertion';
+
+function verifyFillEvents(
+    takerAddress: string,
+    order: Order,
+    receipt: TransactionReceiptWithDecodedLogs,
+    deployment: DeploymentManager,
+): void {
+    // Ensure that the fill event was correct.
+    verifyEvents<ExchangeFillEventArgs>(
+        receipt,
+        [
+            {
+                makerAddress: order.makerAddress,
+                feeRecipientAddress: order.feeRecipientAddress,
+                makerAssetData: order.makerAssetData,
+                takerAssetData: order.takerAssetData,
+                makerFeeAssetData: order.makerFeeAssetData,
+                takerFeeAssetData: order.takerFeeAssetData,
+                orderHash: orderHashUtils.getOrderHashHex(order),
+                takerAddress,
+                senderAddress: takerAddress,
+                makerAssetFilledAmount: order.makerAssetAmount,
+                takerAssetFilledAmount: order.takerAssetAmount,
+                makerFeePaid: constants.ZERO_AMOUNT,
+                takerFeePaid: constants.ZERO_AMOUNT,
+                protocolFeePaid: DeploymentManager.protocolFee,
+            },
+        ],
+        ExchangeEvents.Fill,
+    );
+
+    // Ensure that the transfer events were correctly emitted.
+    verifyEvents<ERC20TokenTransferEventArgs>(
+        receipt,
+        [
+            {
+                _from: takerAddress,
+                _to: order.makerAddress,
+                _value: order.takerAssetAmount,
+            },
+            {
+                _from: order.makerAddress,
+                _to: takerAddress,
+                _value: order.makerAssetAmount,
+            },
+            {
+                _from: takerAddress,
+                _to: deployment.staking.stakingProxy.address,
+                _value: DeploymentManager.protocolFee,
+            },
+        ],
+        ERC20TokenEvents.Transfer,
+    );
+}
+
+/**
+ * A function assertion that verifies that a complete and valid fill succeeded and emitted the correct logs.
+ */
+/* tslint:disable:no-unnecessary-type-assertion */
+export function validFillOrderCompleteFillAssertion(
+    deployment: DeploymentManager,
+): FunctionAssertion<[Order, BigNumber, string], {}, FillResults> {
+    const exchange = deployment.exchange;
+
+    return new FunctionAssertion<[Order, BigNumber, string], {}, FillResults>(exchange.fillOrder.bind(exchange), {
```

Comment:
We had a lifetime issue where the `self` and `this` variables were dropping important wrapper functions (like "_strictEncodeArgs").

Reply:
Thinking about this a little more, we need to re-engineer `FunctionAssertion` such that users don't have to worry about lifetime issues when using the framework. I don't think that it's high-pri, but the best way to accomplish this is probably to pass in the wrapper and the function name of the wrapper function that we care about. A hackier solution would also be to let `FunctionAssertion` "own" the contract wrapper as well as the contract wrapper function. This would ensure that the two objects have the same lifetime, but it's a little icky.

What do you think would be nicer? I can probably put this change into this PR.

Reply:
Yeah the first sounds nicer to me, but we can probably punt on it for now